File: two_pointers/count_number_substrings.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function template for C++

class Solution {
  long long solve(const string &s, int k) {
    int n = s.size();
    int j = 0;
    int c[26] = {0};
    int total = 0;
    long long ans = 0;
    for (int i = 0; i < n; ++i) {
      if (c[s[i] - 'a']++ == 0) {
        ++total;
      }

      while (total > k) {
        if (--c[s[j] - 'a'] == 0) {
          --total;
        }
        ++j;
      }
      ans += i - j + 1;
    }
    return ans;
  }

public:
  long long int substrCount(string s, int k) {
    // code here.
    if (k == 1)
      return solve(s, 1);
    return solve(s, k) - solve(s, k - 1);
  }
};

//{ Driver Code Starts.

int main() {
  int t;
  cin >> t;
  while (t--) {
    string s;
    cin >> s;
    int k;
    cin >> k;
    Solution ob;
    cout << ob.substrCount(s, k) << endl;
  }
}
// } Driver Code Ends

File: two_pointers/count_number_substrings.py
# User function Template for python3


class Solution:
    def count(self, s, k):
        cnt = [0] * 26
        unique = 0
        j = 0
        ans = 0
        n = len(s)
        for i in range(n):
            idx = ord(s[i]) - ord("a")
            cnt[idx] += 1
            if cnt[idx] == 1:
                unique += 1
            while unique > k:
                jdx = ord(s[j]) - ord("a")
                cnt[jdx] -= 1
                if cnt[jdx] == 0:
                    unique -= 1
                j += 1
            ans += j

        return ans

    def substrCount(self, s, k):
        if k == 0:
            return self.count(s, 1)
        return self.count(s, k - 1) - self.count(s, k)


# {
# Driver Code Starts
# Initial Template for Python 3

t = int(input())
for tc in range(t):
    s = input()
    k = int(input())
    ob = Solution()
    print(ob.substrCount(s, k))
# } Driver Code Ends
